### PR TITLE
[SPARK-22114][ML][MLLIB]add epsilon for LDA

### PR DESCRIPTION
**Pull request by mpjlu**
## What changes were proposed in this pull request?

The current convergence condition of OnlineLDAOptimizer is:
while(meanGammaChange > 1e-3)
The condition is critical for the performance and accuracy of LDA. 
We should keep this configurable, like it is in Vowpal Vabbit: 
https://github.com/JohnLangford/vowpal_wabbit/blob/430f69453bc4876a39351fba1f18771bdbdb7122/vowpalwabbit/lda_core.cc :638

## How was this patch tested?

The existing UT

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAOptimizer.scala [hunk 1]
```diff
@@ -573,7 +584,8 @@ private[clustering] object OnlineLDAOptimizer {
       expElogbeta: BDM[Double],
       alpha: breeze.linalg.Vector[Double],
       gammaShape: Double,
-      k: Int): (BDV[Double], BDM[Double], List[Int]) = {
+      k: Int,
+      epsilon: Double = 1e-3): (BDV[Double], BDM[Double], List[Int]) = {
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAOptimizer.scala]: This duplicates the default value.

Comment (mpjlu) [reply mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAOptimizer.scala]: This is for some other code,  like in LDAModel, which calls this function and does not need to set epsilon.

**Review by WeichenXu123 — COMMENTED** [review]
Does it need to check `epsilon > 0` when set ?

**Comment by mpjlu** [comment]
Not check is also ok, user should know epsilon > 0

**Comment by WeichenXu123** [comment]
OK. But I'm afraid accidently negative value cause the code run into dead loop. Maybe adding `assert` is better.

**Comment by mpjlu** [comment]
OK, I will change it to assert.

**Comment by mpjlu** [comment]
Because epsilon is Double,  negative value should not cause the code run into dead loop. All other setting in LDA using require for check or no check. Should we use assert only for this change?

**Comment by WeichenXu123** [comment]
Why not possible ? `meanGammaChange` is always positive. The loop will become `while (true) {...}`
and should use `require` instead of `assert`

**Comment by mpjlu** [comment]
Sorry, I got wrong. 
So you think assert is better here? now we use require.

**Comment by WeichenXu123** [comment]
`require` is better i think

**Comment by mpjlu** [comment]
Ok, thanks. we don't need to change the code here.

**Comment by mgaido91** [comment]
As far as I know, mllib is not maintained anymore. You should add the `epsilon` parameter to the new ml API.

**Comment by mpjlu** [comment]
For LDA, the implementation is in mllib, ml calls mllib.

**Comment by mgaido91** [comment]
I know, but you should add the parameter to the ml API, otherwise the change is pointless.

**Comment by mpjlu** [comment]
Thanks @mgaido91 , I will update the ML api, maybe also python and java API.

File: mllib/src/main/scala/org/apache/spark/ml/clustering/LDA.scala [hunk 2]
```diff
@@ -224,6 +224,20 @@ private[clustering] trait LDAParams extends Params with HasFeaturesCol with HasM
   /**
    * For Online optimizer only: [[optimizer]] = "online".
    *
+   * @group expertParam
+   */
+  @Since("2.3.0")
+  final val epsilon = new DoubleParam(this, "epsilon", "(For online optimizer)" +
+    " A (positive) learning parameter that controls the convergence of variational inference.",
+    ParamValidators.gt(0))
```
Comment (mgaido91) [inline mllib/src/main/scala/org/apache/spark/ml/clustering/LDA.scala]: what about using `0.0` to remark that this is a Double?

Comment (mpjlu) [reply mllib/src/main/scala/org/apache/spark/ml/clustering/LDA.scala]: Yes, use 0.0 is good. Because other Double in this class is using 0, to keep the same code style, I also use 0 here.
It is ok to change all 0 to 0.0.

**Comment by WeichenXu123** [comment]
+1 for updating ML API.

File: mllib/src/main/scala/org/apache/spark/ml/clustering/LDA.scala [hunk 3]
```diff
@@ -224,6 +224,20 @@ private[clustering] trait LDAParams extends Params with HasFeaturesCol with HasM
   /**
    * For Online optimizer only: [[optimizer]] = "online".
    *
+   * @group expertParam
```
Comment (hhbyyh) [inline mllib/src/main/scala/org/apache/spark/ml/clustering/LDA.scala]: parameter comments.

File: mllib/src/main/scala/org/apache/spark/ml/clustering/LDA.scala [hunk 4]
```diff
@@ -224,6 +224,20 @@ private[clustering] trait LDAParams extends Params with HasFeaturesCol with HasM
...
+   */
+  @Since("2.3.0")
+  final val epsilon = new DoubleParam(this, "epsilon", "(For online optimizer)" +
+    " A (positive) learning parameter that controls the convergence of variational inference.",
```
Comment (hhbyyh) [inline mllib/src/main/scala/org/apache/spark/ml/clustering/LDA.scala]: The parameter introduction here cannot really help a user without knowledge of LDA implementation. Please add more description for the effect if user want to tune the parameter, such like "Smaller value will lead to higher accuracy with the cost of more iterations."

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAOptimizer.scala [hunk 5]
```diff
@@ -322,6 +326,13 @@ final class OnlineLDAOptimizer extends LDAOptimizer {
     this
   }
 
+  @Since("2.3.0")
+  def setEpsilon(epsilon: Double): this.type = {
+    require(epsilon> 0, s"LDA epsilon must be positive, but was set to $epsilon")
```
Comment (hhbyyh) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAOptimizer.scala]: space after epsilon.

**Review by hhbyyh — COMMENTED** [review]
Thanks for the PR. Generally looks good. 

In other algorithms, the parameter that controls convergence is usually called "tol". But epsilon is acceptable to me here. 

Please also update the unit tests for parameter check and serialization. 

Please also create jira for python.

**Comment by mpjlu** [comment]
Thanks, @hhbyyh.
 I will create a JIRA for python API

File: mllib/src/test/scala/org/apache/spark/ml/clustering/LDASuite.scala [hunk 6]
```diff
@@ -119,6 +121,8 @@ class LDASuite extends SparkFunSuite with MLlibTestSparkContext with DefaultRead
     assert(lda.getLearningDecay === 0.53)
     lda.setLearningOffset(1027)
     assert(lda.getLearningOffset === 1027)
+    lda.setEpsilon(1e-3)
```
Comment (mgaido91) [inline mllib/src/test/scala/org/apache/spark/ml/clustering/LDASuite.scala]: here you should test a value different from the default, otherwise this tests nothing.

Comment (mpjlu) [reply mllib/src/test/scala/org/apache/spark/ml/clustering/LDASuite.scala]: Thanks @mgaido91 , I will update the code.

Comment (mgaido91) [reply mllib/src/test/scala/org/apache/spark/ml/clustering/LDASuite.scala]: Thanks.

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAOptimizer.scala (hunk 1 above)
Comment (mgaido91) [reply mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAOptimizer.scala]: shouldn't epsilon be set in LDAModel as well to reflect the value which was used while training the model?

Comment (mpjlu) [reply mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAOptimizer.scala]: Maybe not need to set, we can add epsilon as a parameter in the model to save it. Because most of training parameters are not in the model, so we don't add epsilon to the model here. Thanks.

**Review by hhbyyh — COMMENTED** [review]
I just checked the LDA API in scikit.  http://scikit-learn.org/stable/modules/generated/sklearn.decomposition.LatentDirichletAllocation.html 

1. We may need to support other tolerance in the future. Suggest to use `convergenceTolerance` or `convergeTol` for this param name, to be more intuitive and flexible. Sorry to overthrow my comment earlier.

2. From ML API, this param ('epsilon`) will be inherited by LDAModel, and IMO it should make an impact to the topicInference process. We can also add a setter for the param in LocalLDAModel (need unit test). API in MLlib does not need to be changed unless necessary.

Both of my suggestions require quite some code change. Feel free to object or wait for other opinions.

File: mllib/src/main/scala/org/apache/spark/ml/clustering/LDA.scala [hunk 7]
```diff
@@ -224,6 +224,24 @@ private[clustering] trait LDAParams extends Params with HasFeaturesCol with HasM
   /**
    * For Online optimizer only: [[optimizer]] = "online".
    *
+   * A (positive) learning parameter that controls the convergence of variational inference.
+   * Smaller value will lead to more accuracy model and longer training time.
```
[the remark this answers is not in the document]
Comment (mpjlu) [reply mllib/src/main/scala/org/apache/spark/ml/clustering/LDA.scala]: Thanks, I will update the doc.

**Comment by mpjlu** [comment]
For the comments about change the name of epsilon and add setter in localLADModel, we have agreed not to change it now after some offline discussion. 
Because epsilon doesn't control model convergence directly, and some other LDA implementations like Vowpal Vabbit also uses this name. 
Because there are many parameters in LDA, epsilon is just one of them, now there is no setter for any of them. If we need to add setter of them, we maybe add them together in another PR. 
Thanks. @hhbyyh @jkbradley

**Comment by mpjlu** [comment]
Any comments for this?

**Comment by mpjlu** [comment]
Because I don't have the environment to continue this work, I will close it. Thanks.
